internal/config/variables/variables.go

```diff
@@ -0,0 +1,531 @@
+package variables
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/ext/typeexpr"
+	"github.com/hashicorp/hcl/v2/gohcl"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	hcljson "github.com/hashicorp/hcl/v2/json"
+	pb "github.com/hashicorp/waypoint/internal/server/gen"
+	"github.com/zclconf/go-cty/cty"
+	"github.com/zclconf/go-cty/cty/convert"
+	"github.com/zclconf/go-cty/cty/gocty"
+)
+
+var (
+	// Variable value sources
+	// listed in descending precedence order for ease of reference
+	sourceCLI     = "cli"
+	sourceFile    = "file"
+	sourceEnv     = "env"
+	sourceVCS     = "vcs"
+	sourceServer  = "server"
+	sourceDefault = "default"
+
+	// Prefix for collecting variable values from environment variables
+	varEnvPrefix = "WP_VAR_"
+
+	// The attributes we expect to see in variable blocks
+	// Future expansion here could include `sensitive`, `validations`, etc
+	variableBlockSchema = &hcl.BodySchema{
+		Attributes: []hcl.AttributeSchema{
+			{
+				Name: "default",
+			},
+			{
+				Name: "type",
+			},
+			{
+				Name: "description",
+			},
+		},
+	}
+)
+
+// Variable stores a parsed variable definition from the waypoint.hcl
+type Variable struct {
+	Name string
+
+	// The default value in the variable definition
+	Default *InputValue
+
+	// Cty Type of the variable. If the default value or a collected value is
+	// not of this type nor can be converted to this type an error diagnostic
+	// will show up. This allows us to assume that values are valid.
+	//
+	// When a default value - and no type - is passed into the variable
+	// declaration, the type of the default variable will be used.
+	Type cty.Type
+
+	// Description of the variable
+	Description string
+
+	// The location of the variable definition block in the waypoint.hcl
+	Range hcl.Range
+}
+
+type HclVariable struct {
+	Name        string         `hcl:",label"`
+	Default     cty.Value      `hcl:"default,optional"`
+	Type        hcl.Expression `hcl:"type,optional"`
+	Description string         `hcl:"description,optional"`
+}
+
+// InputValues are used to store values collected from various sources.
+// Values are added to the map in precedence order, and then used to
+// create the final map of cty.Values for config hcl context evaluation.
+type InputValues map[string]*InputValue
```

Comment:
This is a later comment after I read the rest of this PR, I want to bikeshed this again and note if we want to name this just `Values` and rename `InputValue` to just `Value`. 

I originally read `InputValue` and thought "what other type of value is there?" And figured I'd run into one but didn't. Is there a reason for the additional qualifier?

Reply:
I named it this way under the assumption we might expand variable types later and have other types of values. But there is no immediate reason to name it so specifically, and so simpler is good for now. I'll name it Values/Value.